### DISCUSSION TITLE
Initializing mesh with data

### DISCUSSION THREAD

**Discussion by Rahul-JOON** [pyvista/pyvista#4619] [Q&A] [answered]
Thanks for the library, Great job team!
I am trying to initialize relevant attributes and methods for an imported stl/obj mesh. May someone help me with some pyvista API methods that can help me to do that; the attributes/methods I want to have is:

Attributes:
1. orientation of mesh; an array for orthogonal vectors of axes that represents the mesh' orientation in 3d space, that preferably updates itself.

Methods:
1. Method for translating the mesh, if given a position vector; translate the faces, face centers, and mesh center.
2. Method for reorienting the mesh, if given a rotation axis (unit vector) and an angle; rotate the faces, face centers, face normals, and mesh center.
3. Method for rescaling the mesh, if given a scale value; scale the faces, face centers, and mesh scale.

**Comment by MatthewFlamm** [comment, marked as answer]
Methods to look into https://docs.pyvista.org/version/stable/api/core/_autosummary/pyvista.DataSet.translate.html#pyvista.DataSet.translate,
https://docs.pyvista.org/version/stable/api/core/_autosummary/pyvista.DataSet.rotate_vector.html#pyvista.DataSet.rotate_vector, and https://docs.pyvista.org/version/stable/api/core/_autosummary/pyvista.DataSet.scale.html#pyvista.DataSet.scale.  Also https://docs.pyvista.org/version/stable/api/core/_autosummary/pyvista.DataSetFilters.transform.html#pyvista.DataSetFilters.transform.

What do you mean by orientation of the mesh?  This would be a 4x4 matrix that is relative to the starting position/rotation after multiple transformations?  I don't think this is available out of the box.  If not, to add an attribute to the mesh that tracks this as methods like `rotate_vector` are called, you can create a subclass of your DataSet type, e.g. 'class MyPolyData(pv.PolyData):`.  Then you will need to create methods for transformations that update this attribute before/after you do the actual transformation using one of the existing methods above.

**Reply by Rahul-JOON** [reply]
Thanks, @MatthewFlamm for the concise and accurate answer.
And regarding the orientation attribute, I was thinking of a general np.ndarray with dimensions as : (3 spatial coordinates, 3 vectors) and if it was possible to track it with relative angle changes as you mentioned or a center of mass orientation tracking capability; but I take it as it is not available as such.
Nevertheless, I have used the native pyvista.Cube() as a tracker of the orientation of the main mesh; I do so by rotating the cube every time along with the main mesh and then extracting orthogonal face normal vectors, this is working fine for now.
Anything else you want to add?
